### DISCUSSION TITLE
How to work with EntityRuler? Id does not return me any results

### DISCUSSION THREAD

**Discussion by aleksandar-devedzic** [explosion/spaCy#7585] [Help: Coding & Implementations] [answered]
I want to make model that will recognise organisations names.
I have added more than 3000 names of the organisations like this:

```
patterns = []
for organisation in organisations_list:
    patterns.append({"label": "ORG", "pattern": organisation.strip()})
```

Then I created a blank model:
```

ruler = EntityRuler(nlp)
ruler.add_patterns(patterns)
nlp = spacy.blank("en")
nlp.add_pipe('entity_ruler')
```
And then, I have tested it like this:

```
for full_text in list_of_texts:
    doc = nlp(full_text)
    print(doc.ents.text, doc.ents.label_)
```

And it does not recognises anything (even if Im testing it in a sentence that has exact name of the company).
What am I doing wrong?
I have noticed that It works if I do it like this:

```
ruler = EntityRuler(nlp)
ruler.add_patterns(patterns)
nlp = spacy.load("en_core_web_trf")
nlp.add_pipe('entity_ruler', before = 'tagger')
#if i do print(nlp.pipeline) i can see entity_ruler added before tager.
```
But then I do not know if it works because of my entity_ruler or because of the pre trained model. I have tested it on 20 example texts and it gives me the same results with entity_ruler and without it.

What am I doing wrong?

**Comment by polm** [comment, marked as answer]
[Answered on Stack Overflow](https://stackoverflow.com/questions/66825806/pythons-spacy-entityruler-does-not-return-me-any-results/66828940#66828940), but you're not adding the ruler correctly. 

Do this instead:

```
nlp = spacy.blank("en")
ruler = nlp.add_pipe("entity_ruler")
ruler.add_patterns(patterns)
```

Your current code creates two separate and unrelated EntityRuler objects, and you're adding your patterns to the one that isn't in the pipeline.

**Reply by aleksandar-devedzic** [reply]
Do I have enough examples for training, is 3000 enough?
Because Im starting with empty model, after adding those 3000 or more names to the model, is it possible for model to recognise those entity names if the spelling or letter capitalisations are different?
Is it possible to recognise company/organisation names that are not in these 3000 examples?

**Reply by polm** [reply]
If you have 3000 annotated examples (complete sentences) then that is enough to start training a statistical model, but it sounds like you just have the names of 3000 companies, which can't be used as training data. I suggest you read the ["Should I use rules or train a model?"](https://spacy.io/usage/rule-based-matching#rules-vs-model) section in the docs.

Also note that the EntityRuler is **not a model**, it is a rule-based component for detecting entities. You can't train it.

> Because Im starting with empty model, after adding those 3000 or more names to the model, is it possible for model to recognise those entity names if the spelling or letter capitalisations are different?

You can tell the EntityRuler to match on the lower case form of a word so that changes in capitalization don't matter. You could also tell it to use lemmas so that some kinds of changes in spelling don't matter ("Apples" would be the same as "Apple"). There isn't an example of this in the documentation yet, but look at how the config works in [the EntityRuler docs](https://spacy.io/api/entityruler#config), specifically the `phrase_matcher_attr`. 

> Is it possible to recognise company/organisation names that are not in these 3000 examples?

Not with the EntityRuler.

This is one of the tradeoffs of rule-based vs machine learning methods. With the rule-based EntityRuler, you don't need training data, so you can get started right away. But in order to recognize things not in your list you'll need training data to train a model.
